internal/provider/resource_variable.go

```diff
@@ -0,0 +1,282 @@
+			},
+			"hcl": {
+				Type:                types.BoolType,
+				MarkdownDescription: "Whether this variable has an HCL value.",
+				Description:         "Whether this variable has an HCL value.",
+				Required:            true,
+			},
+			"key": {
+				Type:                types.StringType,
+				MarkdownDescription: "This variable's key (within its namespace).",
+				Description:         "This variable's key (within its namespace).",
+				Required:            true,
+			},
+			"value": {
+				Type:                types.StringType,
+				MarkdownDescription: "This variable's value.",
+				Description:         "This variable's value.",
+				Optional:            true,
+			},
+		},
+	}, nil
+}
+
+// Configure lets the provider implement the ResourceWithConfigure interface.
+func (t *variableResource) Configure(_ context.Context,
+	req resource.ConfigureRequest, _ *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+	t.client = req.ProviderData.(*tharsis.Client)
+}
+
+func (t *variableResource) Create(ctx context.Context,
+	req resource.CreateRequest, resp *resource.CreateResponse) {
+
+	// Retrieve values from namespace variable.
+	var variable VariableModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &variable)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Create the namespace variable.
+	created, err := t.client.Variable.CreateVariable(ctx,
+		&ttypes.CreateNamespaceVariableInput{
+			NamespacePath: variable.NamespacePath.ValueString(),
+			Category:      ttypes.VariableCategory(variable.Category.ValueString()),
+			HCL:           variable.Hcl.ValueBool(),
+			Key:           variable.Key.ValueString(),
+			Value:         variable.Value.ValueString(),
+		})
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error creating namespace variable",
+			err.Error(),
+		)
+		return
+	}
+
+	// Map the response body to the schema and update the plan with the computed attribute values.
+	// Because the schema uses the Set type rather than the List type, make sure to set all fields.
+	t.copyVariable(*created, &variable)
+
+	// Set the response state to the fully-populated plan, whether or not there is an error.
+	resp.Diagnostics.Append(resp.State.Set(ctx, variable)...)
+}
+
+func (t *variableResource) Read(ctx context.Context,
+	req resource.ReadRequest, resp *resource.ReadResponse) {
+
+	// Get the current state.
+	var state VariableModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get the namespace variable from Tharsis.
+	found, err := t.client.Variable.GetVariable(ctx, &ttypes.GetNamespaceVariableInput{
+		ID: state.ID.ValueString(),
+	})
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error reading namespace variable",
+			err.Error(),
+		)
+		return
+	}
+
+	if found == nil {
+		// Handle the case that the namespace variable no longer exists if that fact is reported by returning nil.
+		resp.State.RemoveResource(ctx)
```

Comment:
This is missing a return statement (there may be other places this is missing as well)